src/lib.rs

```diff
@@ -22,23 +23,31 @@ pub struct Matches {
 
 impl fmt::Debug for Matches {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{} has {} matche(s)", self.path, self.lines.len())
+        write!(f, "{} has {} match(es)", self.path, self.lines.len())
     }
 }
 
+#[tracing::instrument]
 pub async fn find_matches(path: &Utf8PathBuf, pattern: &str) -> Result<Matches> {
     let path_as_str = path.to_string();
-    let file = File::open(path).with_context(|| format!("Error opening {}", &path_as_str))?;
+    let file = File::open(path)
+        .await
+        .with_context(|| format!("Error opening {}", &path_as_str))?;
     let reader = BufReader::new(file);
     let mut matches = Matches {
         path: path.to_string(),
         lines: Vec::new(),
     };
 
-    for (idx, result) in reader.lines().enumerate() {
-        let number = idx + 1;
-        let line = result
-            .with_context(|| format!("Error reading line {} from {}", number, &path_as_str))?;
+    let mut number = 0;
+    let mut lines = reader.lines();
+
+    while let Some(line) = lines
+        .next_line()
+        .await
+        .with_context(|| format!("Error reading line {} from {}", number, &path_as_str))?
+    {
+        number += 1;
```

Comment:
This one was the largest change, because I couldn't make `.enumerate()` work =(

But! Hopefully the code is still readable. The main change is using a `while let` instead of a for loop, because we can call `.await` here to yield to other calculations while waiting for the next line.

Reply:
Nice! This looks like the sort of Rust-style approach I'd never get to by myself : ) Thank you!